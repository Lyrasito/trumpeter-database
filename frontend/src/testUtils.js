const fakePlayer = () => ({
  id: 1,
  name: "Fake Player",
  city: "Fake City",
  startYear: 1234,
  endYear: 5678,
});

const fakePlayers = () => [
  {
    id: 1,
    name: "Fake Player",
    city: "Fake City",
    startYear: 1234,
    endYear: 5678,
  },
  {
    id: 2,
    name: "Fake Player 2",
    city: "Fake City 2",
    startYear: 2345,
    endYear: 6789,
  },
];

const fakeAlbums = () => [
  {
    id: 1,
    title: "Fake Album 1",
    year: 1234,
    genre: "fakeGenre1",
    player_id: 1,
  },
  {
    id: 2,
    title: "Fake Album 2",
    year: 1234,
    genre: "fakeGenre2",
    player_id: 1,
  },
];

const fakeAlbum = () => ({
  id: 1,
  title: "Fake Album 1",
  year: 1234,
  genre: "fakeGenre1",
  player_id: 1,
});

const fakeGenres = () => ["fakeGenre1", "fakeGenre2"];

export { fakeAlbum, fakeAlbums, fakePlayer, fakeGenres, fakePlayers };
