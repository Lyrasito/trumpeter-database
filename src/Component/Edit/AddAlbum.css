.album-input {
  box-sizing: border-box;
  height: 63px;
  width: 200px;
  border: 3px solid #666666;
  border-radius: 30px;
  background-color: #ffffff;
  display: block;
  font-size: 24px;
}

.newAlbum {
  display: flex;
  justify-content: space-around;
  width: 1300px;
}

.submit {
  height: 60px;
  width: 162px;
  border-radius: 30px;
  background-color: #ffc600;
  cursor: pointer;
  color: #000000;
  font-family: "WorkSans";
  font-size: 24px;
  font-weight: 600;
  letter-spacing: 0;
  line-height: 35px;
}

@media only screen and (min-width: 1301px) {
  .newAlbum {
    width: 90%;
  }
}
