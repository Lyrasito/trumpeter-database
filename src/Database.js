const fetch = require("node-fetch");

const baseUrl = "http://localhost:4000/api";

const Database = {
  async getPlayers() {
    const url = `${baseUrl}/players`;
    const response = await fetch(url);
    const jsonResponse = await response.json();
    const players = await jsonResponse.players.map((player) => ({
      id: player.id,
      name: player.name,
      city: player.city,
      startYear: player.start_year,
      endYear: player.end_year,
      image: player.image,
    }));
    return players;
  },

  async searchPlayers(city, year, genre) {
    let url = "";
    if (city && !year && !genre) {
      url = `${baseUrl}/players/search?city=${city}`;
    } else if (year && !city && !genre) {
      url = `${baseUrl}/players/search?year=${year}`;
    } else if (genre && !city && !year) {
      url = `${baseUrl}/players/search?genre=${genre}`;
    } else if (city && year && !genre) {
      url = `${baseUrl}/players/search?city=${city}&year=${year}`;
    } else if (city && genre && !year) {
      url = `${baseUrl}/players/search?city=${city}&genre=${genre}`;
    } else if (genre && year && !city) {
      url = `${baseUrl}/players/search?genre=${genre}&year=${year}`;
    } else if (city && year && genre) {
      url = `${baseUrl}/players/search?city=${city}&year=${year}&genre=${genre}`;
    }
    const response = await fetch(url);
    const jsonResponse = await response.json();
    const players = await jsonResponse.players.map((player) => ({
      id: player.id,
      name: player.name,
      city: player.city,
      startYear: player.start_year,
      endYear: player.end_year,
      image: player.image,
    }));
    return players;
  },

  async searchByName(name) {
    const url = `${baseUrl}/players/search?name=${name}`;
    const response = await fetch(url);
    const jsonResponse = await response.json();
    const players = await jsonResponse.players.map((player) => ({
      id: player.id,
      name: player.name,
      city: player.city,
      startYear: player.start_year,
      endYear: player.end_year,
      image: player.image,
    }));
    return players;
  },

  async getPlayerAlbums(player) {
    const url = `${baseUrl}/players/${player.id}/albums`;
    const response = await fetch(url);
    const jsonResponse = await response.json();
    const albums = await jsonResponse.albums.map((album) => ({
      id: album.id,
      title: album.title,
      year: album.year,
    }));
    return albums;
  },

  async getPlayerGenres(player) {
    const url = `${baseUrl}/players/${player.id}/albums/genres`;
    const response = await fetch(url);
    const jsonResponse = await response.json();
    return jsonResponse.genre;
  },

  async addPlayer(name, city, startYear, endYear) {
    const url = `${baseUrl}/players`;
    const response = await fetch(url, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        player: {
          name: name,
          city: city,
          startYear: startYear,
          endYear: endYear,
        },
      }),
    });

    const jsonResponse = await response.json();
    return jsonResponse.player;
  },

  async addAlbum(playerId, title, year, genre) {
    const url = `${baseUrl}/players/${playerId}/albums`;
    const response = await fetch(url, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        album: {
          title: title,
          year: year,
          genre: genre,
          playerId: playerId,
        },
      }),
    });
    const jsonResponse = await response.json();
    console.log(jsonResponse);
  },
};

module.exports = Database;

/*  */
