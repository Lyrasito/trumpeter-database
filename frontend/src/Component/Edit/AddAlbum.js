import React from "react";
import Database from "../../Database";
import "./AddAlbum.css";

class AddAlbum extends React.Component {
  state = {
    playerList: [],
    playerId: null,
    currentPlayer: null,
    title: "",
    year: "",
    genre: "",
    newAlbum: {},
    error: null,
  };

  componentDidMount() {
    this.getAllPlayers();
  }

  getAllPlayers = async () => {
    const allPlayers = await Database.getPlayers();
    this.setState({
      playerList: allPlayers,
    });
  };
  getPlayerId = async (event) => {
    const playerId = event.target.value;
    const response = await Database.getPlayerById(playerId);
    this.setState({
      playerId: playerId,
      currentPlayer: response,
      newAlbum: {},
    });
  };
  handleChange = (event) => {
    const { name, value, type } = event.target;
    const val = type === "number" ? parseFloat(value) : value;
    this.setState({
      [name]: val,
    });
  };

  createAlbum = async () => {
    try {
      const response = await Database.addAlbum(
        this.state.playerId,
        this.state.title,
        this.state.year,
        this.state.genre
      );
      this.setState({ newAlbum: response, title: "", year: "", genre: "" });
    } catch (err) {
      this.setState({ error: err.message, newAlbum: {} });
    }
  };
  renderAlbum = () => {
    if (this.state.newAlbum.title) {
      return (
        <h3 className="message" data-test="success-message">
          You have added {this.state.newAlbum.title} to the library of{" "}
          {this.state.currentPlayer.name}!
        </h3>
      );
    }
  };
  renderError = () => {
    if (this.state.error) {
      return (
        <h4 className="message" data-test="error-message">
          {this.state.error}
        </h4>
      );
    }
  };
  render() {
    return (
      <div>
        <h3 className="message">Add an album to a player's library!</h3>
        <div className="newAlbum">
          <label>
            Select Player
            <select
              className="album-input"
              id="players"
              data-test="playerId"
              onChange={this.getPlayerId}
              defaultValue="Select Player"
            >
              <option value=""></option>
              {this.state.playerList.map((player) => {
                return (
                  <option
                    value={player.id}
                    key={player.id}
                    onSelect={this.getPlayerId}
                  >
                    {player.name}
                  </option>
                );
              })}
            </select>
          </label>
          <label>
            Album title
            <input
              className="album-input"
              id="createTitle"
              name="title"
              data-test="title"
              onChange={this.handleChange}
              value={this.state.title}
            ></input>
          </label>
          <label>
            Release year
            <input
              className="album-input"
              id="createYear"
              name="year"
              data-test="year"
              type="number"
              onChange={this.handleChange}
              value={this.state.year}
            ></input>
          </label>
          <label>
            Album genre
            <input
              className="album-input"
              id="createGenre"
              name="genre"
              data-test="genre"
              onChange={this.handleChange}
              value={this.state.genre}
            ></input>
          </label>

          <button type="submit" className="submit" onClick={this.createAlbum}>
            Add Album
          </button>
        </div>
        {this.renderAlbum()}
        {this.renderError()}
      </div>
    );
  }
}

export default AddAlbum;
