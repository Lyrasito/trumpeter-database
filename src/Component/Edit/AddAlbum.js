import React from "react";
import Database from "../../Database";
import "./AddAlbum.css";

class AddAlbum extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      playerList: [],
      playerId: null,
      title: "",
      year: null,
      genre: "",
      newAlbum: {},
    };
    this.getAllPlayers = this.getAllPlayers.bind(this);
    this.getPlayerId = this.getPlayerId.bind(this);
    this.getTitle = this.getTitle.bind(this);
    this.getYear = this.getYear.bind(this);
    this.getGenre = this.getGenre.bind(this);
    this.createAlbum = this.createAlbum.bind(this);
  }

  componentWillMount() {
    this.getAllPlayers();
  }

  async getAllPlayers() {
    const allPlayers = await Database.getPlayers();
    this.setState({
      playerList: allPlayers,
    });
  }
  getPlayerId(event) {
    const playerId = event.target.value;
    this.setState({ playerId: playerId });
    console.log(playerId, this.state.playerId);
  }
  getTitle(event) {
    this.setState({
      title: event.target.value,
    });
  }
  getYear(event) {
    this.setState({
      year: event.target.value,
    });
  }
  getGenre(event) {
    this.setState({
      genre: event.target.value,
    });
  }
  async createAlbum() {
    const response = await Database.addAlbum(
      this.state.playerId,
      this.state.title,
      this.state.year,
      this.state.genre
    );
    this.setState({ newAlbum: response });
    console.log(response);
  }
  render() {
    return (
      <div>
        <h3>Add an album to a player's library!</h3>
        <div className="newAlbum">
          <select
            className="album-input"
            id="players"
            onChange={this.getPlayerId}
            defaultValue="Select Player"
          >
            <option value="">Select Player</option>
            {this.state.playerList.map((player) => {
              return (
                <option
                  value={player.id}
                  key={player.id}
                  onSelect={this.getPlayerId}
                >
                  {player.name}
                </option>
              );
            })}
          </select>

          <input
            className="album-input"
            id="createTitle"
            placeholder="Title"
            onChange={this.getTitle}
          ></input>
          <input
            className="album-input"
            id="createYear"
            placeholder="Year"
            onChange={this.getYear}
          ></input>
          <input
            className="album-input"
            id="createGenre"
            placeholder="Genre"
            onChange={this.getGenre}
          ></input>
          <button type="submit" className="submit" onClick={this.createAlbum}>
            Create
          </button>
        </div>
      </div>
    );
  }
}

export default AddAlbum;
