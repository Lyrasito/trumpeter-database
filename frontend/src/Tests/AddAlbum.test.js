import AddAlbum from "../Component/Edit/AddAlbum";
import React from "react";
import toJSON from "enzyme-to-json";
import { mount, shallow } from "enzyme";
import wait from "waait";
import { fakeAlbum, fakePlayers } from "../testUtils";
import Database from "../Database";

const players = fakePlayers();
const album = fakeAlbum();

Database.getPlayers = jest.fn().mockResolvedValue(players);
Database.getPlayerById = jest.fn().mockResolvedValue(players[0]);
Database.addAlbum = jest
  .fn()
  .mockResolvedValueOnce(album)
  .mockImplementation(() => {
    throw new Error("Please fill out all fields");
  });

describe("<AddAlbum />", () => {
  it("renders properly", () => {
    const wrapper = shallow(<AddAlbum />);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
  it("adds an album to a player's library", async () => {
    const wrapper = mount(<AddAlbum />);
    const playerId = wrapper.find("[data-test='playerId']");
    const title = wrapper.find("[data-test='title']");
    const year = wrapper.find("[data-test='year']");
    const genre = wrapper.find("[data-test='genre']");
    const button = wrapper.find(".submit");

    playerId.simulate("change", { target: { value: players[0].id } });
    title.instance().value = album.title;
    title.simulate("change");
    year.instance().value = album.year;
    year.simulate("change");
    genre.instance().value = album.genre;
    genre.simulate("change");

    await wait();
    wrapper.update();

    button.simulate("click");

    await wait();
    wrapper.update();

    const message = wrapper.find("[data-test='success-message']");
    expect(message.text()).toBe(
      `You have added ${album.title} to the library of ${players[0].name}!`
    );
    expect(Database.addAlbum).toHaveBeenCalledWith(
      players[0].id,
      album.title,
      album.year,
      album.genre
    );
  });
  it("renders an error if called with insufficient input", async () => {
    const wrapper = mount(<AddAlbum />);
    const playerId = wrapper.find("[data-test='playerId']");
    playerId.simulate("change", { target: { value: players[0].id } });
    const button = wrapper.find(".submit");
    button.simulate("click");

    await wait();
    wrapper.update();
    const message = wrapper.find("[data-test='error-message']");
    expect(message.text()).toBe("Please fill out all fields");
  });
});
