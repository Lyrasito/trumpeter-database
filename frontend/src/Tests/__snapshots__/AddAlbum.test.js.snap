// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddAlbum /> renders properly 1`] = `
<div>
  <h3
    className="message"
  >
    Add an album to a player's library!
  </h3>
  <div
    className="newAlbum"
  >
    <label>
      Select Player
      <select
        className="album-input"
        data-test="playerId"
        defaultValue="Select Player"
        id="players"
        onChange={[Function]}
      >
        <option
          value=""
        />
      </select>
    </label>
    <label>
      Album title
      <input
        className="album-input"
        data-test="title"
        id="createTitle"
        name="title"
        onChange={[Function]}
        value=""
      />
    </label>
    <label>
      Release year
      <input
        className="album-input"
        data-test="year"
        id="createYear"
        name="year"
        onChange={[Function]}
        type="number"
        value=""
      />
    </label>
    <label>
      Album genre
      <input
        className="album-input"
        data-test="genre"
        id="createGenre"
        name="genre"
        onChange={[Function]}
        value=""
      />
    </label>
    <button
      className="submit"
      onClick={[Function]}
      type="submit"
    >
      Add Album
    </button>
  </div>
</div>
`;
